emails/views.py

```diff
@@ -1784,84 +1739,293 @@ def _handle_complaint(message_json: AWS_SNSMessageJSON) -> HttpResponse:
+    complained_recipients: list[tuple[str, dict[str, Any]]]
+    from_addresses: list[str]
+    subtype: str
+    user_agent: str
+    feedback_type: str
+
+
+def _get_complaint_data(message_json: AWS_SNSMessageJSON) -> RawComplaintData:
+    """Extract complaint data from a complaint notification"""
+    complaint = message_json["complaint"]
+
+    T = TypeVar("T")
+
+    def get_or_log(
+        key: str, source: dict[str, T], data_type: type[T]
+    ) -> tuple[T, bool]:
+        """Get a value from a dictionary, or log if not found"""
+        if key in source:
+            return source[key], True
+        logger.error(
+            "_get_complaint_data: Unexpected message",
+            extra={"missing_key": key, "found_keys": ",".join(sorted(source.keys()))},
+        )
+        return data_type(), False
+
+    raw_recipients, has_cr = get_or_log("complainedRecipients", complaint, list)
+    complained_recipients = []
+    no_entries = True
+    for entry in raw_recipients:
+        no_entries = False
+        raw_email_address, has_email = get_or_log("emailAddress", entry, str)
+        if has_email:
+            email_address = parseaddr(raw_email_address)[1]
+            extra = {
+                key: value for key, value in entry.items() if key != "emailAddress"
+            }
+            complained_recipients.append((email_address, extra))
+    if has_cr and no_entries:
+        logger.error("_get_complaint_data: Empty complainedRecipients")
+
+    mail, has_mail = get_or_log("mail", message_json, dict)
+    if has_mail:
+        commonHeaders, has_ch = get_or_log("commonHeaders", mail, dict)
+    else:
+        commonHeaders, has_ch = {}, False
+    if has_ch:
+        raw_from_addresses, _ = get_or_log("from", commonHeaders, list)
+    else:
+        raw_from_addresses = []
+    from_addresses = [parseaddr(addr)[1] for addr in raw_from_addresses]
+
+    feedback_type, _ = get_or_log("complaintFeedbackType", complaint, str)
+
+    # Only present when destination is on account suppression list
+    subtype = complaint.get("complaintSubType", "")
+    # Only present for feedback reports
+    user_agent = complaint.get("userAgent", "")
+
+    return RawComplaintData(
+        complained_recipients, from_addresses, subtype, user_agent, feedback_type
+    )
+
+
+class Complainer(TypedDict):
+    user: User
+    found_in: Literal["complained_recipients", "from_header", "all"]
+    domain: str
+    extra: dict[str, Any] | None
+    masks: list[RelayAddress | DomainAddress]
+
+
+def _gather_complainers(
+    complaint_data: RawComplaintData,
+) -> tuple[list[Complainer], int]:
+    """Fetch Relay Users and masks from the complaint data"""
+
+    users: dict[int, Complainer] = {}
+    unknown_complainer_count = 0
+    for email_address, extra_data in complaint_data.complained_recipients:
+        local, domain = email_address.split("@", 1)
+
+        # For developer mode complaint simulation, swap with developer's email
```

Comment:
nit (non-blocking): this code seems specific to AWS complaint simulator, but not necessarily to the "developer mode" controlled by the mask label?

If it is, can we link from here to a doc/comment about the "developer mode"?

```suggestion
        # For AWS complaint simulation, swap with developer's email
```